File: training/cpp/src/de/sots/cellarsandcentaurs/domain/model/Creature.cpp
#include "Creature.hpp"

#include "../../application/Application.hpp"
#include "Dice.hpp"

namespace de::sots::cellarsandcentaurs::domain::model {

Creature::Creature(CreatureId id)
    : Creature(std::move(id), application::CreatureFacade::STANDARD_CREATURE_TYPE) {}

Creature::Creature(CreatureId id, CreatureType type)
    : id_{std::move(id)}, type_{type} {}

int Creature::attack() const {
    return rollD20().value;
}

bool Creature::isAlive() const {
    return hitPoints_.has_value() && hitPoints_->getCurrent() > 0;
}

}

File: sample-projects/objectivec/CellarsAndCentaurs/Domain/Model/Creature.m
#import "Creature.h"
#import "CreatureId.h"
#import "ArmorClass.h"
#import "HitPoints.h"
#import "Speed.h"
#import "Dice.h"
#import "CellarsAndCentaurs.h"

@interface Creature ()

@property (nonatomic, assign) NSInteger initiative;

@end

@implementation Creature

@synthesize speeds = _speeds;

- (instancetype)initWithId:(CreatureId *)creatureId {
    return [self initWithId:creatureId type:CreatureFacade.standardCreatureType];
}

- (instancetype)initWithId:(CreatureId *)creatureId type:(CreatureType)type {
    self = [super init];
    if (self) {
        _creatureId = [creatureId copy];
        _type = type;
        _speeds = @{};
    }
    return self;
}

- (Speed *)speedOfType:(SpeedType)speedType {
    return self.speeds[@(speedType)];
}

#pragma mark - Fightable

- (NSInteger)rollInitiative {
    self.initiative = rollD20();
    return self.initiative;
}

- (void)takeDamage:(NSInteger)damage {
    [self.hitPoints loseHitPoints:damage];
}

@end
